### PR TITLE
Improvements for Trace Computation

### LINKED ISSUE #336: Improvements for Trace Computation
At the moment, the `partial_trace` (and, hence, the `trace`) function in the package are calculated in a straight-forward fashion. As a result, the performance of the operation scales with the number of paths in the decision diagram as opposed to the number of nodes. Introducing a compute table as for the other decision diagram operations fixes this problem. A possible benchmark to test this is to simply create the matrix of an n-qubit Hadamard transform and compute its trace. This matrix has `n` nodes, but `4^n` paths.

Furthermore, it might be possible to implement an improved `trace` function, that does not just call the partial trace and eliminates every variable, and instead just sums up all the diagonal entries.
It might even be beneficial to compute `trace(...)/2^n` in order to normalise the result to the range `[0, 1]`. Instead of computing `trace(...)` and then dividing by `2^n` (which induces numerical problems), the computed trace can be divided by two in each of the levels in order to continuously normalise it throughout the computation.

### PR DESCRIPTION
**Pull request by TeWas**
## Description

This PR introduces improvements for the trace computation: 

1. Adds a compute table for storing results of eliminated nodes during (partial) trace computation
- this allows the full trace computation to scale linearly with the number of nodes
- however, the partial trace computation still scales with the number of paths in the DD if the bottom qubits are to be eliminated:  non-eliminated nodes cannot be stored in the compute table, as this would prevent their proper elimination in subsequent trace calls
2. Normalizes the result

Fixes #336 

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: include/mqt-core/dd/Package.hpp [hunk 1]
```diff
@@ -1111,7 +1118,7 @@ template <class Config> class Package {
     auto tmp0 = conjugateTranspose(measZeroDd);
     auto tmp1 = multiply(e, densityFromMatrixEdge(tmp0), false);
     auto tmp2 = multiply(densityFromMatrixEdge(measZeroDd), tmp1, true);
-    auto densityMatrixTrace = trace(tmp2, nrQubits);
+    auto densityMatrixTrace = trace(tmp2, nrQubits) * std::pow(2, nrQubits);
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]: Technically, I think we could get rid of this by performing the normalization in the `trace` method based on whether we operate on density matrices or matrices.
For density matrices, it always holds that the trace is 1 (It's a summation of probabilities).
For unitary matrices, the largest trace is the one of the identity, which is $2^n$.

So, if we only divide by two for each level whenever we are dealing with matrices, we can guarantee (and commnicate in the docstring) that the `trace` method will return a value in the interval $[0,1]$.

File: include/mqt-core/dd/Package.hpp [hunk 2]
```diff
@@ -276,10 +279,12 @@ template <class Config> class Package {
       vectorInnerProduct.clear();
       vectorKronecker.clear();
       matrixKronecker.clear();
+      matrixTrace.clear();
       stochasticNoiseOperationCache.clear();
       densityAdd.clear();
       densityDensityMultiplication.clear();
       densityNoise.clear();
+      densityTrace.clear();
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp:291]: I might be mistaken, but I think you do not need to clear these two tables here.
Both tables take only a `Node*` as an input and return a `CachedEdge`, i.e., none of the operands contains a complex number from the respective table. Consequently, it should be fine to leave them as they are if only complex numbers have been collected.

File: include/mqt-core/dd/Package.hpp [hunk 3]
```diff
@@ -260,12 +261,14 @@ template <class Config> class Package {
...
       conjugateMatrixTranspose.clear();
       matrixKronecker.clear();
+      matrixTrace.clear();
       matrixVectorMultiplication.clear();
       matrixMatrixMultiplication.clear();
       stochasticNoiseOperationCache.clear();
       densityAdd.clear();
       densityDensityMultiplication.clear();
       densityNoise.clear();
+      densityTrace.clear();
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]: This is technically not really part of this PR, but it is kind of triggered by the change here:
It might make sense to separate the `||` into separate `if`'s here given the fact that no tables operating on density matrices need to be cleared whenever matrix nodes were collected and vice versa.

Would you mind adjusting that (including the top-level comment before the `if`)?

File: include/mqt-core/dd/Package.hpp [hunk 4]
```diff
@@ -1985,16 +2012,17 @@ template <class Config> class Package {
       return CachedEdge<Node>::zero();
     }
 
-    if (std::none_of(eliminate.begin(), eliminate.end(),
-                     [](bool v) { return v; })) {
+    // If a is the identity matrix or there is nothing left to eliminate (e.g. a
+    // is a terminal) then simply return a
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]:
```suggestion
    // If `a` is the identity matrix or there is nothing left to eliminate,
    // then simply return `a`
```
The terminal itself is considered the identity, but it could very well be that someone requests the partial trace and only wants to trace out a single qubit.

File: include/mqt-core/dd/Package.hpp [hunk 5]
```diff
@@ -1985,16 +2012,17 @@ template <class Config> class Package {
...
+    if (a.isIdentity() || std::none_of(eliminate.begin(), eliminate.end(),
+                                       [](bool v) { return v; })) {
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]: Now that I think about it: 
Isn't the check here a bit too pessimistic? In the sense that it always checks the full `eliminated` vector to be none, while it would technically be enough for a stopping condition that none of the qubits from index `0` (very bottom of the DD) up to the current `level` (or `level-1`; needs to be checked) should be eliminated.

Say someone want to trace out only the top-most qubit in a three-qubit system. Then they would pass `eliminate={false, false, true}`, which would never trigger the condition here and would always trigger a recursive computation. Although, after the first level, one could technically stop, because no further levels are eliminated.
It should be possible to realize that by simply exchanging the `eliminate.end()` with the appropriate `eliminate.begin() + x` (where `x` is an index based on the current `level`).
This should drastically improve the partial trace performance.
Maybe you can also add a test for that.

File: include/mqt-core/dd/Package.hpp [hunk 6]
```diff
@@ -2003,6 +2031,8 @@ template <class Config> class Package {
       auto r = add2(trace(a.p->e[0], eliminate, level - 1, elims),
                     trace(a.p->e[3], eliminate, level - 1, elims), v - 1);
 
+      r.w = r.w / 2.0;
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]: This line could maybe use a comment stating that this is where we apply the normalization.

Regarding the other comment from above: this is also the place where the normalization would only be applied to matrices and not density matrices.

File: include/mqt-core/dd/Package.hpp [hunk 7]
```diff
@@ -2019,6 +2049,7 @@ template <class Config> class Package {
                                 eliminate.begin(), eliminate.end(), true)) -
                             alreadyEliminated));
     auto r = makeDDNode(adjustedV, edge);
+    // nodes that were not eliminated are not added to the compute table
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]: Given that such nodes are not put in the compute table, I am wondering whether one could slightly optimize the compute table lookup at the beginning of this function. In the sense, that no lookup needs to be performed if it is clear from the start that it cannot be successful.

File: test/dd/test_package.cpp [hunk 8]
```diff
@@ -285,21 +285,86 @@ TEST(DDPackageTest, IdentityTrace) {
...
+  auto swapGate2 = dd2->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
   auto ptr = dd->partialTrace(swapGate, {true, false});
-  EXPECT_EQ(ptr.w * ptr.w, 1.);
+  auto fullTrace = dd->trace(ptr, 1);
+  auto fullTraceOriginal = dd2->trace(swapGate2, 2);
+  EXPECT_EQ(dd::RealNumber::val(ptr.w.r), 0.5);
+  // Check that successively tracing out subsystems is the same as computing the
+  // full trace from the beginning
+  EXPECT_EQ(fullTrace.r, fullTraceOriginal.r);
+}
```
Comment (burgholzer) [inline test/dd/test_package.cpp]: Really only a minor comment, but I think you technically do not need to allocate two DD packages here (and two SWAP gates). It should be enough to allocate one package and to create the SWAP DD once.
The rest of the test should "simply work"

File: test/dd/test_package.cpp [hunk 9]
```diff
@@ -285,21 +285,86 @@ TEST(DDPackageTest, IdentityTrace) {
...
+
+TEST(DDPackageTest, PartialTraceKeepInnerQubits) {
+  // Check that the partial trace computation is correct when tracing out the
+  // outer qubits only. This test shows that we should avoid storing
+  // non-eliminated nodes in the compute table, as this would prevent their
+  // proper elimination in subsequent trace calls.
+
+  const std::size_t numQubits = 8;
+  auto dd = std::make_unique<dd::Package<>>(numQubits);
+  auto dd2 = std::make_unique<dd::Package<>>(numQubits);
```
Comment (burgholzer) [inline test/dd/test_package.cpp]: Same as above goes here. You should be able to reduce the number of packages to one.

File: test/dd/test_package.cpp [hunk 10]
```diff
@@ -285,21 +285,86 @@ TEST(DDPackageTest, IdentityTrace) {
...
+    auto hKron = hGate;
+    for (std::size_t i = 0; i < numQubits - 1; ++i) {
+      hKron = dd->kronecker(hKron, hGate, 1);
+    }
+    dd->trace(hKron, numQubits);
+    auto stats = computeTable.getStats();
+    auto lookups = stats.lookups;
+    auto hits = stats.hits;
+    ASSERT_EQ(lookups, 2 * numQubits - 1);
+    ASSERT_EQ(hits, numQubits - 1);
```
Comment (burgholzer) [inline test/dd/test_package.cpp]:
```suggestion
    const auto& stats = computeTable.getStats();
    ASSERT_EQ(stats.lookups, 2 * numQubits - 1);
    ASSERT_EQ(stats.hits, numQubits - 1);
```
Just a tiny bit more compact.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Many thanks for your work on this and for the clear code comments. These are highly appreciated!
I looked through all the changes here and you'll find my comments inline. Some are just minor suggestions that you can quickly incorporate. Some might take a little bit of work, but I think it shouldn't be too bad.

File: include/mqt-core/dd/Package.hpp (hunk 2 above)
Comment (TeWas) [reply include/mqt-core/dd/Package.hpp:291]: Oh yes, indeed. It seems like this is also the case for some other tables, such as vectorKronecker and matrixKronecker, where only the final returned edge weight is stored in cUniqueTable. I will check the others as well!

File: test/dd/test_package.cpp (hunk 9 above)
Comment (TeWas) [reply test/dd/test_package.cpp]: The issue here is that `partialTrace `and `trace `cannot share the same compute table. Depending on the execution order, one computation will produce incorrect results. For example, if `fullTraceOriginal `is computed first, the partial trace computation will be incorrect because all qubits were eliminated and thus the nodes were stored in the compute table. But then, during the partial trace computation, for the top node, the result from the compute table will be used, even though some lower-level qubits should have been eliminated.

File: include/mqt-core/dd/Package.hpp (hunk 2 above)
Comment (burgholzer) [reply include/mqt-core/dd/Package.hpp:291]: That could very well be the case. Some of these have only recently been refactored to just take a `Node` instead of an `Edge`.
Thanks for looking into it!

File: test/dd/test_package.cpp (hunk 9 above)
Comment (burgholzer) [reply test/dd/test_package.cpp]: I think you can work around that, by only inserting nodes into the compute table if all nodes below that node are eliminated as well. That disambiguates the two use cases. And while it limits the caching for the partial trace a little, it should allow to use both functions within the same package instance.

In the end, it would be very undesirable if a user would get unexpected results because they called the trace method followed by the partial trace. We should make sure that both of them can work in parallel, even if that means that one of them will perform slightly worse.

File: include/mqt-core/dd/Package.hpp [hunk 11]
```diff
@@ -1985,24 +2026,51 @@ template <class Config> class Package {
...
     if (eliminate[v]) {
+      // Lookup nodes marked for elimination in the compute table if all
+      // lower level qubits are eliminated as well: if the trace has already
+      // been computed, return the result
+      auto& computeTable = getTraceComputeTable<Node>();
+      if (std::all_of(
+              eliminate.begin(),
+              eliminate.begin() +
+                  static_cast<std::vector<bool>::difference_type>(level),
+              [](bool e) { return e; })) {
```
Comment (burgholzer) [inline include/mqt-core/dd/Package.hpp]: Just a quick side comment (besides the currently failing unittest): I think you can extract this computation to avoid repeating it at the end of this block.
```c++
const auto computeFullTrace = std::all_of(
              eliminate.begin(),
              eliminate.begin() +
                  static_cast<std::vector<bool>::difference_type>(level),
              [](bool e) { return e; });
```
or something similar.

Comment (TeWas) [reply include/mqt-core/dd/Package.hpp]: Done, thanks!

File: test/dd/test_package.cpp (hunk 9 above)
Comment (TeWas) [reply test/dd/test_package.cpp]: Yes, so the cases are disambiguated now by checking before the lookup and insertion into the table that the lower-level qubits are eliminated as well.

File: include/mqt-core/dd/Package.hpp (hunk 2 above)
Comment (burgholzer) [reply include/mqt-core/dd/Package.hpp:291]: As discussed offline, I transferred this to a separate issue. See #633

**Review by burgholzer — APPROVED** [review]
Many thanks for all your work on this. Especially, for writing thorough tests! ✅ 

I just pushed some finishing touches, which are mostly stylistic/minor. After CI is all green, this is good to go 🚀
